File: linked_list/138_copy_list_with_random_pointer.cc
#include <unordered_map>

class Node {
 public:
  int val;
  Node* next;
  Node* random;

  Node(int _val) {
    val = _val;
    next = nullptr;
    random = nullptr;
  }
};

// 2024/01/07: 自己写出迭代+哈希表版本.
Node* copyRandomList(Node* head) {
  std::unordered_map<Node*, Node*> hash_map;
  Node* curr_node = head;
  while (curr_node) {
    Node* temp_node = new Node(curr_node->val);
    hash_map[curr_node] = temp_node;
    curr_node = curr_node->next;
  }
  curr_node = head;
  while (curr_node) {
    hash_map[curr_node]->next = hash_map[curr_node->next];
    hash_map[curr_node]->random = hash_map[curr_node->random];
    curr_node = curr_node->next;
  }
  return hash_map[head];
}

// 2024/01/07: 直接抄答案, 回溯 + 哈希表
std::unordered_map<Node*, Node*> cached_node_;
Node* CopyRandomList2(Node* head) {
  if (head == nullptr) {
    return nullptr;
  }
  if (!cached_node_.count(head)) {
    Node* head_new = new Node(head->val);
    cached_node_[head] = head_new;
    head_new->next = CopyRandomList2(head->next);
    head_new->random = CopyRandomList2(head->random);
  }
  return cached_node_[head];
}

// 2024/01/07: 直接抄答案, 迭代 + 节点拆分
Node* CopyRandomList(Node* head) {
  if (head == nullptr) {
    return nullptr;
  }
  for (Node* node = head; node != nullptr; node = node->next->next) {
    Node* node_new = new Node(node->val);
    node_new->next = node->next;
    node->next = node_new;
  }
  for (Node* node = head; node != nullptr; node = node->next->next) {
    Node* node_new = node->next;
    node_new->random = node->random ? node->random->next : nullptr;
  }
  Node* head_new = head->next;
  for (Node* node = head; node != nullptr; node = node->next) {
    Node* node_new = node->next;
    node->next = node->next->next;
    node_new->next = node_new->next ? node_new->next->next : nullptr;
  }
  return head_new;
}

File: linked_list/138_copy_list_with_random_pointer.py
from linked_list.list_node import ListNode, create_list_node, print_list_node
from typing import Optional, List


class Node:
    def __init__(self, x: int, next: "Node" = None, random: "Node" = None):
        self.val = int(x)
        self.next = next
        self.random = random


def create_node(nums: List) -> Optional[Node]:
    dummy = Node(0)
    curr = dummy

    for num in nums:
        curr.next = Node(num)
        curr = curr.next

    return dummy.next


def create_random_node(head: List[List]) -> Optional[None]:
    nums = [item[0] for item in head]
    random = [item[1] for item in head]
    length = len(nums)
    node_list = create_node(nums)

    hash_map = {}
    curr = node_list
    for i in range(length):
        if curr is None:
            break
        hash_map[i] = curr
        curr = curr.next

    for i in range(length):
        hash_map[i].random = hash_map[random[i]] if random[i] is not None else None

    return node_list


# 2026/05/14: 自己写, 但是用了 AI 进行 Debug.
def copy_random_list(head: "Optional[Node]") -> "Optional[Node]":
    hash_map = {}
    dummy_new_head = Node(0)
    curr = head
    new_curr = dummy_new_head
    while curr is not None:
        new_node = Node(curr.val)
        new_curr.next = new_node
        hash_map[curr] = new_node
        curr = curr.next
        new_curr = new_curr.next

    for key, value in hash_map.items():
        value.random = hash_map[key.random] if key.random is not None else None

    return dummy_new_head.next


head = [[7, None], [13, 0], [11, 4], [10, 2], [1, 0]]
head = create_random_node(head)
new_head = copy_random_list(head)
new_head
